src/WebJobs.Script/Binding/GeneralScriptBindingProvider.cs

```diff
@@ -71,7 +72,7 @@ internal static Type GetRequestedType(ScriptBindingContext context)
             {
                 // arrays are supported for both trigger input as well
                 // as output bindings
-                type = type.MakeArrayType();
+                type = context.SupportsDeferredBinding ? typeof(IEnumerable<ParameterBindingData>) : type.MakeArrayType();
```

Comment:
@mathewc and @fabiocav looking for feedback on the change in this line. 

`MakeArrayType()` converts an object to its Array type - for example `string` becomes `string[]` and `ParameterBindingData` becomes `ParameterBindingData[]`. This array type is not able to bind to anything in extension code - [link](https://github.com/Azure/azure-sdk-for-net/blob/126ed7e63f99e305a962ac43613d55a25ab5fbfe/sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/Config/BlobsExtensionConfigProvider.cs#L110)

We wanted Array to bind to below line but it gives error that `expected IEnumerable<> but found array []`

```csharp
rule.AddOpenConverter<MultiBlobContext, IEnumerable<BlobCollectionType>>(typeof(BlobCollectionConverter<>), this);
```

Reply:
Yeah, something seems wrong here - we shouldn't have to special case `IEnumerable<ParameterBindingData>` above. What OOP bindings does this array code currently work for? For example, the EventHub binding supports binding to EventData[] (e.g. [here](IEnumerable<ParameterBindingData>) in code, and [here](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-event-hubs-trigger?tabs=in-process%2Cfunctionsv2%2Cextensionv5&pivots=programming-language-javascript#example) in doc). We should verify array works for EventHub OOP and figure out the difference. Perhaps we just need to update the blob extension to support binding to arrays. At the WebJobs SDK level, we have enumerable tests for the trigger (e.g. [here](https://github.com/Azure/azure-sdk-for-net/blob/4162f6fa2445b2127468b9cfd080f01c9da88eba/sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/samples/readmes/BlobExtensionSamples.cs#L261)) - does a test binding to `BlobClient[]` work as well?

Reply:
On EventHub's side, below are the files doing specific handling for Array type `EventData[]` -

- [EventHubTriggerBindingStrategy](https://github.com/Azure/azure-sdk-for-net/blob/abed7d9c80073dde103af4a3b1f1daa7bccb9cdb/sdk/eventhub/Microsoft.Azure.WebJobs.Extensions.EventHubs/src/Triggers/EventHubTriggerBindingStrategy.cs#L40) - ```public EventData[] BindMultiple(EventHubTriggerInput value)```
- [EventHubTriggerInput](https://github.com/Azure/azure-sdk-for-net/blob/abed7d9c80073dde103af4a3b1f1daa7bccb9cdb/sdk/eventhub/Microsoft.Azure.WebJobs.Extensions.EventHubs/src/Triggers/EventHubTriggerInput.cs#L20)
- [EventHubExtensionConfigProvider](https://github.com/Azure/azure-sdk-for-net/blob/abed7d9c80073dde103af4a3b1f1daa7bccb9cdb/sdk/eventhub/Microsoft.Azure.WebJobs.Extensions.EventHubs/src/Config/EventHubExtensionConfigProvider.cs#L66)


On Blob extensions side, currently `BlobClient[]` does not work because there is no rule in [BlobsExtensionConfigProvider](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/Config/BlobsExtensionConfigProvider.cs) that can bind to array type (error - `expected IEnumerable<> but found array []`). Also, their [documentation](https://github.com/Azure/azure-sdk-for-net/blob/4162f6fa2445b2127468b9cfd080f01c9da88eba/sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/samples/readmes/BlobExtensionSamples.cs#L261) says that `IEnumerable<BlobClient>` works and they have a rule to bind to that in [BlobsExtensionConfigProvider](https://github.com/Azure/azure-sdk-for-net/blob/abed7d9c80073dde103af4a3b1f1daa7bccb9cdb/sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/Config/BlobsExtensionConfigProvider.cs#L102)

```csharp
rule.AddOpenConverter<MultiBlobContext, IEnumerable<BlobCollectionType>>(typeof(BlobCollectionConverter<>), this);
```

So, in summary, Host is converting to `Array type` but blob extension supports only `Enumerable type` while EventHub supports `array type`. Based on this, we have 2 options now -

1. Either, support `IEnumerable<>` in Host which I did above. One advantage of this approach is that we need only a single line change to support `IEnumerable<>`.
2. Or, support `Array` in Blob extensions. For this, I have prototyped changes below - 
          - On Host side, keep the line 74 above as it is: `type = type.MakeArrayType();`. 
          - On Blob Extensions side - [full changes in draft mode](https://github.com/surgupta-msft/azure-sdk-for-net/compare/sdk-bindings/support-reference-type...surgupta-msft:azure-sdk-for-net:sdk-bindings/collection-changes?expand=1)


``` csharp
// Add new rule for Array []. Rule for IEnumerable is already there in this class.
rule.AddOpenConverter<MultiBlobContext, BlobCollectionType[]>(typeof(BlobCollectionArrayConverter<>), this);

// Add new converter for Array. Converter for IEnumerable is already there in this class.
 private class BlobCollectionArrayConverter<T> : IAsyncConverter<MultiBlobContext, T[]>
```

With protoypted changes, option 2 is causing lot of code duplication between `IEnumberable<BlobClient>` and `BlobClient[]` as there are 2 separate converters for them, but if option-2 makes more sense then we can discuss how to minimize duplication.

Reply:
Updated code based on discussion offline. Also added changes in [blob extension PR](https://github.com/Azure/azure-sdk-for-net/pull/31757/commits/58c9eecd0d1f8525816ad24fd7590708cd6f36b9) to support array type